cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java

```diff
@@ -117,7 +147,7 @@ public DynamicDatasetCache(SystemDatasetInstantiator instantiator,
     // The internal of the dataset instantiate may switch the context class loader to a different one when necessary
     ClassLoader currentClassLoader = ClassLoaders.setContextClassLoader(getClass().getClassLoader());
     try {
-      return getDataset(new DatasetCacheKey(name, dsArguments), bypass);
+      return getDataset(new DatasetCacheKey(name, dsArguments, accessType), bypass);
```

Comment:
To avoid caching the dataset per accessType would mean that we need to record lineage outside of the dataset cache (not just upon dataset retrieval, but also upon access of a cached dataset).

Also, we still need some way to propagate the accessType to the load function of the cache, which isn't clear how that will be done.